stock_picking_location_check/models/stock_picking.py

```diff
@@ -0,0 +1,51 @@
+# Copyright 2024 Quartile (https://www.quartile.co)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+
+from odoo import _, fields, models
+from odoo.exceptions import UserError
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    allow_location_inconsistency = fields.Boolean(
+        copy=False,
+        help="If enabled, no error is raised for location inconsistency between "
+        "picking and its move lines at the time of validation.",
+    )
+
+    def _get_child_location_ids(self, location):
+        return self.env["stock.location"].search(
+            [("id", "child_of", location.id), ("usage", "!=", "view")]
+        )
+
+    def _check_location_consistency(self, pick_location, line_locations):
+        self.ensure_one()
+        if not set(line_locations).issubset(
+            self._get_child_location_ids(pick_location)
+        ):
+            raise UserError(
+                _(
+                    "A move line location is not related to that of the picking: %s",
+                    self.name,
+                )
+            )
+
+    def _action_done(self):
+        for pick in self:
+            if pick.allow_location_inconsistency:
+                continue
+            line_source_locations = [
+                line.location_id
+                for line in pick.move_line_ids
+                if not getattr(line.move_id, "is_subcontract", False)
```

Comment:
Why is `is_subcontract` field here as there is no dependency on `mrp_subcontracting`. If you have problem with that, a glue module is necessary.

Reply:
If `mrp_subtcontracting` was installed in the odoo environment, we want to filter out these subcontracted move lines because they are coming from `subcontracting location` and their location and picking's location will never be the same. Then, I think `getattr` is enough for this and we don't need to depend on `mrp_subcontracting`.